File: 0002_simple_db_multi_segments/cpp/simple_db_multi_segments.cpp
#include "simple_db_multi_segments.h"
#include <chrono>
#include <ctime>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <set>

const char *IsADirectoryError::what() const noexcept {
  return "Not a directory, or the directory isn't formatted correctly";
}

Index::Index(const std::string &segment_name)
    : segment_name(segment_name), cursor(0) {}

void Index::add_next(const std::string &line, const std::string &key) {
  std::string actual_key = key;
  if (key.empty()) {
    size_t pos = line.find(',');
    actual_key = line.substr(0, pos);
  }
  size_t length = line.length();
  idx_map[actual_key] = {cursor, length};
  cursor += length;
}

std::pair<size_t, size_t> Index::get(const std::string &key) const {
  auto it = idx_map.find(key);
  if (it != idx_map.end()) {
    return it->second;
  }
  throw std::runtime_error("Key not found");
}

const std::string &Index::get_segment_name() const { return segment_name; }

SimpleDbMultiSegments::SimpleDbMultiSegments(const std::string &dbname,
                                             size_t segment_bytes_threshold)
    : dbname(dbname), indexes_loaded(false),
      segment_bytes_threshold(std::max(segment_bytes_threshold, size_t(1))) {
  check_db_directory();
  load_indexes();
}

void SimpleDbMultiSegments::set(const std::string &key,
                                const nlohmann::json &json_dict) {
  if (indexes.empty() ||
      indexes.back()->get_cursor() >= segment_bytes_threshold) {
    auto index = std::make_shared<Index>(
        dbname + "/segment_" +
        std::to_string(get_epoch_time_in_microseconds()) + ".db");
    indexes.push_back(index);
  }
  auto index = indexes.back();
  std::string line = key + "," + json_dict.dump() + "\n";
  index->add_next(line, key);
  std::ofstream file(index->get_segment_name(), std::ios::app);
  file << line;
}

nlohmann::json SimpleDbMultiSegments::get(const std::string &key) {
  for (auto it = indexes.rbegin(); it != indexes.rend(); ++it) {
    try {
      auto [offset, length] = (*it)->get(key);
      std::ifstream file((*it)->get_segment_name());
      file.seekg(offset);
      std::string line(length, '\0');
      file.read(&line[0], length);
      return nlohmann::json::parse(line.substr(line.find(',') + 1));
    } catch (const std::exception &) {
      continue;
    }
  }
  return nullptr;
}

void SimpleDbMultiSegments::compact(size_t new_segment_bytes_threshold) {
  size_t current_segment_bytes_threshold = new_segment_bytes_threshold
                                               ? new_segment_bytes_threshold
                                               : segment_bytes_threshold;

  std::set<std::string> checked_keys;
  std::vector<std::shared_ptr<Index>> new_indexes;
  auto new_index = std::make_shared<Index>(
      dbname + "/segment_" + std::to_string(get_epoch_time_in_microseconds()) +
      ".db");

  for (auto it = indexes.rbegin(); it != indexes.rend(); ++it) {
    for (const auto &[key, _] : (*it)->get_idx_map()) {
      if (checked_keys.find(key) != checked_keys.end())
        continue;
      checked_keys.insert(key);

      if (new_index->get_cursor() >= current_segment_bytes_threshold) {
        new_indexes.push_back(new_index);
        new_index = std::make_shared<Index>(
            dbname + "/segment_" +
            std::to_string(get_epoch_time_in_microseconds()) + ".db");
      }

      nlohmann::json value = get(key);
      std::string line = key + "," + value.dump() + "\n";
      new_index->add_next(line, key);
      std::ofstream file(new_index->get_segment_name(), std::ios::app);
      file << line;
    }
  }

  if (!new_index->get_idx_map().empty()) {
    new_indexes.push_back(new_index);
  }

  indexes = std::move(new_indexes);
}

void SimpleDbMultiSegments::check_db_directory() {
  std::filesystem::path directory(dbname);
  std::filesystem::path marker = directory / ".simple_db_multi_segments_marker";

  if (std::filesystem::exists(directory)) {
    if (std::filesystem::is_directory(directory)) {
      if (!std::filesystem::exists(marker)) {
        throw IsADirectoryError();
      }
    } else {
      throw IsADirectoryError();
    }
  } else {
    std::filesystem::create_directory(directory);
    std::ofstream(marker).close();
  }
}

void SimpleDbMultiSegments::load_index(Index &index) {
  std::ifstream file(index.get_segment_name());
  std::string line;
  while (std::getline(file, line)) {
    index.add_next(line + "\n");
  }
}

void SimpleDbMultiSegments::load_indexes() {
  if (!indexes_loaded) {
    auto directory = std::filesystem::path(dbname);
    for (const auto &entry : std::filesystem::directory_iterator(directory)) {
      if (entry.path().filename().string().find("segment_") == 0) {
        auto index = std::make_shared<Index>(entry.path().string());
        indexes.push_back(index);
        load_index(*index);
      }
    }
    indexes_loaded = true;
  }
}

int64_t SimpleDbMultiSegments::get_epoch_time_in_microseconds() const {
  auto now = std::chrono::system_clock::now();
  auto epoch = now.time_since_epoch();
  auto microseconds =
      std::chrono::duration_cast<std::chrono::microseconds>(epoch).count();
  return microseconds;
}

File: 0002_simple_db_multi_segments/python/simple_db_multi_segments.py
import asyncio
import functools
import json
import time
from pathlib import Path


class IsADirectoryError(Exception):
    pass


class _Index:
    def __init__(self, segment_name):
        self._segment_name = segment_name
        self._idx_map = {}
        self._cursor = 0

    async def add_next(self, line: str, key=None):
        if key is None:
            key, _  = line.split(',', 1)
        length = len(line)
        self._idx_map[key] = (self._cursor, length)
        self._cursor += length

    async def get(self, key):
        return self._idx_map.get(key)



# A segment file has a name pattern of 'segment_{unix epoch microsecond timestamp}.db'
_SEGMENT_FILE_PATTERN = "segment_*.db"


def _check_db_directory(dbname):
    directory = Path(dbname)
    marker = directory / '.simple_db_multi_segments_marker'

    if directory.exists():
        if directory.is_dir():
            if not marker.exists():
                raise IsADirectoryError
        else:
            raise IsADirectoryError
    else:
        directory.mkdir()
        marker.touch()

    return directory


async def _load_index(index):
    with open(index._segment_name, 'r') as f:
        tasks = []
        for line in f:
            tasks.append(asyncio.create_task(index.add_next(line)))
        for task in tasks:
            await task


def _load_indexes(func):
    @functools.wraps(func)
    async def wrapper(self, *args, **kwargs):
        if not self._indexes_loaded:
            directory = _check_db_directory(self.dbname)    

            tasks = []
            matching_files = sorted(directory.glob(_SEGMENT_FILE_PATTERN))
            for file_path in matching_files:
                index = _Index(file_path)
                self._indexes.append(index)
                tasks.append(asyncio.create_task(_load_index(index)))
            for task in tasks:
                await task
            self._indexes_loaded = True

        return await func(self, *args, **kwargs)

    return wrapper


class SimpleDbMultiSegments:
    def __init__(self, dbname='database', segment_bytes_threshold=1024 * 1024):
        self.dbname = dbname
        self._indexes = []
        self._indexes_loaded = False
        self._segment_bytes_threshold = max(segment_bytes_threshold, 1)  # Ensure segment_bytes_threshold is at least 1

    @_load_indexes
    async def set(self, key, json_dict):
        if len(self._indexes) < 1 or self._indexes[-1]._cursor >= self._segment_bytes_threshold:
            index = _Index(f'{self.dbname}/segment_{int(time.time() * 1_000_000)}.db')
            self._indexes.append(index)
        else:
            index = self._indexes[-1]

        line = f'{key},{json.dumps(json_dict)}\n'
        task = asyncio.create_task(index.add_next(line, key=key))
        with open(index._segment_name, 'a') as f:
            f.write(line)
        await task

    @_load_indexes
    async def get(self, key):
        for index in reversed(self._indexes):
            try:
                offset, length = await index.get(key) 
                break
            except TypeError:
                continue
        else:
            return None

        with open(index._segment_name, 'r') as f:
            f.seek(offset)
            return json.loads(f.read(length).split(',', 1)[1].strip())

    @_load_indexes
    async def compact(self, new_segment_bytes_threshold=None):
        if new_segment_bytes_threshold is None:
            segment_bytes_threshold = self._segment_bytes_threshold
        else:
            segment_bytes_threshold = new_segment_bytes_threshold

        checked_keys = set()
        new_indexes = []
        new_index = _Index(f'{self.dbname}/segment_{int(time.time() * 1_000_000)}.db')
        for index in reversed(self._indexes):
            for key in index._idx_map:
                if key in checked_keys:
                    continue
                checked_keys.add(key)

                if new_index._cursor >= segment_bytes_threshold:
                    new_indexes.append(new_index)
                    new_index = _Index(f'{self.dbname}/segment_{int(time.time() * 1_000_000)}.db')

                value = await self.get(key)
                line = f'{key},{json.dumps(value)}\n'
                await new_index.add_next(line, key=key)

                with open(new_index._segment_name, 'a') as f:
                    f.write(line)

        if new_index._idx_map:
            new_indexes.append(new_index)

        self._indexes = new_indexes